File: src/main/kotlin/syntaxlm/SemanticModels.kt
package syntaxlm

import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest
import kotlin.math.abs
import kotlin.math.exp
import kotlin.math.tanh

/** All positions are inclusive start/exclusive end UTF-16 offsets. */
data class SymbolResult(
    val id: String, val documentId: String, val revision: String,
    val start: Int, val end: Int, val name: String, val kind: Int, val confidence: Float,
    val contextEnd: Int? = null,
) {
    init {
        require(id.isNotEmpty() && documentId.isNotEmpty() && revision.isNotEmpty())
        require(start >= 0 && end >= start && kind in 0..7 && confidence in 0f..1f)
        require(contextEnd == null || contextEnd >= end)
    }
}

data class SemanticSnapshot(
    val documentId: String, val revision: String,
    val definitions: List<SymbolResult>, val usages: List<SymbolResult>,
    val definitionLanguages: FloatArray, val usageLanguages: FloatArray,
)

data class DefinitionResolution(
    val usage: SymbolResult, val definition: SymbolResult?, val confidence: Float,
    val candidateProbabilities: FloatArray,
) { val status: String get() = if (definition == null) "unresolved" else "resolved" }

internal fun sha256(bytes: ByteArray): String = MessageDigest.getInstance("SHA-256")
    .digest(bytes).joinToString("") { "%02x".format(it.toInt() and 255) }

/** Minimal strict JSON reader for embedded SYL2 metadata; no Python or JSON library dependency. */
private class MetadataJson(private val text: String) {
    private var at = 0
    private fun space() { while (at < text.length && text[at].isWhitespace()) at++ }
    private fun expect(c: Char) { space(); require(at < text.length && text[at++] == c) { "invalid JSON" } }
    private fun string(): String {
        expect('"')
        val out = StringBuilder()
        while (at < text.length) {
            val c = text[at++]
            if (c == '"') return out.toString()
            if (c == '\\') {
                require(at < text.length)
                when (val escaped = text[at++]) {
                    '"', '\\', '/' -> out.append(escaped)
                    'n' -> out.append('\n'); 'r' -> out.append('\r'); 't' -> out.append('\t')
                    'b' -> out.append('\b'); 'f' -> out.append('\u000c')
                    'u' -> { require(at + 4 <= text.length); out.append(text.substring(at, at + 4).toInt(16).toChar()); at += 4 }
                    else -> error("invalid JSON escape")
                }
            } else { require(c.code >= 32); out.append(c) }
        }
        error("unterminated JSON string")
    }
    private fun value(depth: Int = 0): Any? {
        require(depth <= 32) { "metadata nesting limit" }
        space(); require(at < text.length)
        return when (text[at]) {
            '"' -> string()
            '{' -> {
                at++; val map = linkedMapOf<String, Any?>(); space()
                if (text[at] != '}') while (true) {
                    val key = string(); require(!map.containsKey(key)); expect(':'); map[key] = value(depth + 1); space()
                    if (text[at] != ',') break
                    at++
                }
                expect('}'); map
            }
            '[' -> {
                at++; val list = arrayListOf<Any?>(); space()
                if (text[at] != ']') while (true) {
                    list.add(value(depth + 1)); space(); if (text[at] != ',') break; at++
                }
                expect(']'); list
            }
            else -> {
                val start = at
                while (at < text.length && text[at] !in ",]} \t\r\n") at++
                when (val token = text.substring(start, at)) {
                    "true" -> true; "false" -> false; "null" -> null
                    else -> token.toDouble().also { require(it.isFinite()) }
                }
            }
        }
    }
    @Suppress("UNCHECKED_CAST")
    fun parse(): Map<String, Any?> {
        val result = value() as Map<String, Any?>; space(); require(at == text.length); return result
    }
}

internal data class Tensor(val shape: IntArray, val data: FloatArray)

internal class SemanticWeights(file: File, task: String) {
    val hash: String
    val metadata: Map<String, Any?>
    val tensors: Map<String, Tensor>
    init {
        require(file.length() in 48..16_777_216) { "semantic artifact size limit" }
        val raw = file.readBytes()
        hash = sha256(raw)
        val body = raw.copyOfRange(0, raw.size - 32)
        require(MessageDigest.getInstance("SHA-256").digest(body).contentEquals(raw.copyOfRange(raw.size - 32, raw.size))) { "SYL2 checksum mismatch" }
        val b = ByteBuffer.wrap(body).order(ByteOrder.LITTLE_ENDIAN)
        fun bytes(n: Int): ByteArray { require(n >= 0 && n <= b.remaining()); return ByteArray(n).also { b.get(it) } }
        fun str(): String = String(bytes(b.int), Charsets.UTF_8)
        require(String(bytes(4), Charsets.US_ASCII) == "SYL2" && b.int == 1) { "invalid SYL2 header" }
        metadata = MetadataJson(str()).parse()
        require(metadata["schema"] == "semantic-records-v1" && metadata["taskId"] == task) { "incompatible semantic artifact" }
        require(metadata["featureNames"] == SemanticModelProvider.FEATURE_NAMES) { "feature order mismatch" }
        val result = linkedMapOf<String, Tensor>()
        val count = b.int; require(count in 1..100)
        repeat(count) {
            val name = str(); require(!result.containsKey(name))
            require(b.get().toInt() == 1) { "only float32 tensors supported" }
            val rank = b.get().toInt() and 255; require(rank in 1..4)
            val shape = IntArray(rank) { b.int.also { require(it in 1..65536) } }
            var elements = 1L
            for (dim in shape) { elements *= dim; require(elements <= 4_000_000) }
            require(b.long == elements * 4 && elements * 4 <= b.remaining())
            result[name] = Tensor(shape, FloatArray(elements.toInt()) { b.float.also { require(it.isFinite()) } })
        }
        require(!b.hasRemaining())
        tensors = result
    }
    fun tensor(name: String, vararg dimensions: Int): FloatArray {
        val tensor = tensors.getValue(name)
        require(tensor.shape.contentEquals(dimensions)) { "wrong tensor shape for $name" }
        return tensor.data
    }
    fun trained(): Boolean = (metadata["trainedSteps"] as Number).toInt() > 0
}

internal fun softmax(input: FloatArray): FloatArray {
    val max = input.maxOrNull() ?: return FloatArray(0)
    val out = FloatArray(input.size) { exp((input[it] - max).toDouble()).toFloat() }
    val sum = out.sum()
    return FloatArray(out.size) { out[it] / sum }
}
private fun argmax(a: FloatArray): Int = a.indices.maxByOrNull { a[it] } ?: 0
private fun linear(input: FloatArray, w: FloatArray, bias: FloatArray): FloatArray = FloatArray(bias.size) { row ->
    var sum = bias[row]; val offset = row * input.size
    for (i in input.indices) sum += w[offset + i] * input[i]
    sum
}

/** Mirrors PyTorch GRU r/z/n gate order, including reset-after hidden bias. */
private class GruDirection(weights: SemanticWeights, suffix: String, val inputSize: Int) {
    private val wi = weights.tensor("encoder.weight_ih_$suffix", 192, inputSize)
    private val wh = weights.tensor("encoder.weight_hh_$suffix", 192, 64)
    private val bi = weights.tensor("encoder.bias_ih_$suffix", 192)
    private val bh = weights.tensor("encoder.bias_hh_$suffix", 192)
    fun run(input: Array<FloatArray>, reverse: Boolean): Array<FloatArray> {
        var state = FloatArray(64)
        val output = Array(input.size) { FloatArray(64) }
        for (step in input.indices) {
            val index = if (reverse) input.lastIndex - step else step
            val x = linear(input[index], wi, bi)
            val h = linear(state, wh, bh)
            val next = FloatArray(64) { j ->
                val r = 1.0 / (1.0 + exp(-(x[j] + h[j]).toDouble()))
                val z = 1.0 / (1.0 + exp(-(x[64 + j] + h[64 + j]).toDouble()))
                val n = tanh(x[128 + j] + r * h[128 + j])
                ((1.0 - z) * n + z * state[j]).toFloat()
            }
            state = next; output[index] = next
        }
        return output
    }
}

class SemanticDetector(file: File, val role: String) {
    internal val weights = SemanticWeights(file, role)
    val sequenceBytes = (weights.metadata["sequenceBytes"] as Number).toInt()
    private val embedding = weights.tensor("embedding.weight", 259, 64)
    private val directions = (0..1).map { layer ->
        Pair(GruDirection(weights, "l$layer", if (layer == 0) 64 else 128),
             GruDirection(weights, "l${layer}_reverse", if (layer == 0) 64 else 128))
    }
    private val tw = weights.tensor("occurrence_head.weight", 3, 128)
    private val tb = weights.tensor("occurrence_head.bias", 3)
    private val kw = weights.tensor("kind_head.weight", 8, 128)
    private val kb = weights.tensor("kind_head.bias", 8)
    private val languageCount = (weights.metadata["languages"] as List<*>).size
    val languages: List<String> = (weights.metadata["languages"] as List<*>).map { it as String }
    private val lw = weights.tensor("language_head.weight", languageCount, 128)
    private val lb = weights.tensor("language_head.bias", languageCount)
    init { require(role in listOf("definitions", "usages") && sequenceBytes >= 4) }
    internal fun forward(ids: IntArray): Triple<Array<FloatArray>, Array<FloatArray>, FloatArray> {
        var output = Array(ids.size) { embedding.copyOfRange(ids[it] * 64, ids[it] * 64 + 64) }
        for ((forward, backward) in directions) {
            val f = forward.run(output, false); val b = backward.run(output, true)
            output = Array(ids.size) { f[it] + b[it] }
        }
        val mean = FloatArray(128) { j -> output.sumOf { it[j].toDouble() }.toFloat() / ids.size }
        return Triple(Array(ids.size) { linear(output[it], tw, tb) },
                      Array(ids.size) { linear(output[it], kw, kb) }, linear(mean, lw, lb))
    }
    fun detect(text: String, documentId: String, revision: String): Pair<List<SymbolResult>, FloatArray> {
        require(documentId.isNotEmpty() && revision.isNotEmpty())
        val units = arrayListOf(0); val bytes = arrayListOf(0)
        var unit = 0; var byte = 0
        while (unit < text.length) {
            val cp = text.codePointAt(unit)
            require(cp !in 0xD800..0xDFFF) { "unpaired surrogate in source" }
            unit += Character.charCount(cp)
            byte += when { cp < 128 -> 1; cp < 2048 -> 2; cp < 65536 -> 3; else -> 4 }
            units.add(unit); bytes.add(byte)
        }
        val raw = text.toByteArray(Charsets.UTF_8)
        val chars = units.size - 1
        val tags = IntArray(chars); val confidence = FloatArray(chars)
        val kinds = Array(chars) { FloatArray(8) }; val language = FloatArray(languageCount)
        var a = 0
        while (a < chars) {
            var b = a + 1
            while (b < chars && bytes[b + 1] - bytes[a] <= sequenceBytes) b++
            val input = IntArray(bytes[b] - bytes[a] + 1) { if (it == 0) 256 else raw[bytes[a] + it - 1].toInt() and 255 }
            val (tagLogits, kindLogits, langLogits) = forward(input)
            val lang = softmax(langLogits)
            for (j in lang.indices) language[j] += lang[j] * (b - a)
            for (i in a until b) {
                val index = bytes[i] - bytes[a] + 1
                val probs = softmax(tagLogits[index]); tags[i] = argmax(probs)
                confidence[i] = probs[tags[i]]; kinds[i] = kindLogits[index]
            }
            a = b
        }
        if (chars > 0) for (j in language.indices) language[j] /= chars
        val records = arrayListOf<SymbolResult>(); var start = -1
        for (i in 0..chars) {
            val tag = if (i == chars) 0 else tags[i]
            if (start >= 0 && tag != 2) {
                val k = FloatArray(8); var certainty = 0f
                for (j in start until i) { for (c in k.indices) k[c] += kinds[j][c]; certainty += confidence[j] }
                val first = units[start]; val end = units[i]
                records.add(SymbolResult("$role:$first:$end", documentId, revision, first, end,
                                         text.substring(first, end), argmax(k), certainty / (i - start), text.length))
                start = -1
            }
            if (i < chars && tag != 0 && start < 0) start = i
        }
        return Pair(records, language)
    }
}

class SemanticModelProvider(definitionsFile: File, usagesFile: File, intelligenceFile: File) {
    private val definitions = SemanticDetector(definitionsFile, "definitions")
    private val usages = SemanticDetector(usagesFile, "usages")
    private val intelligence = SemanticWeights(intelligenceFile, "code-intelligence")
    private val hw = intelligence.tensor("hidden.weight", 64, FEATURE_NAMES.size)
    private val hb = intelligence.tensor("hidden.bias", 64)
    private val sw = intelligence.tensor("score.weight", 1, 64)
    private val sb = intelligence.tensor("score.bias", 1)
    val languages: List<String> get() = definitions.languages
    init {
        val dependencies = intelligence.metadata["dependencies"] as Map<*, *>
        require(dependencies["definitions"] == definitions.weights.hash && dependencies["usages"] == usages.weights.hash) {
            "intelligence was trained with different detector artifacts"
        }
        require(definitions.weights.metadata["registryHash"] == usages.weights.metadata["registryHash"] &&
                definitions.weights.metadata["registryHash"] == intelligence.metadata["registryHash"])
        require(definitions.languages == usages.languages && definitions.languages == intelligence.metadata["languages"])
    }
    fun analyze(text: String, documentId: String, cursor: Int? = null): SemanticSnapshot {
        require(definitions.weights.trained() && usages.weights.trained()) { "detectors have no training steps" }
        if (cursor != null) {
            require(cursor in 0..text.length)
            require(cursor == 0 || cursor == text.length || !(text[cursor - 1].isHighSurrogate() && text[cursor].isLowSurrogate()))
        }
        val context = if (cursor == null) text else text.substring(0, cursor)
        val revision = sha256(text.toByteArray(Charsets.UTF_8))
        val (defs, dl) = definitions.detect(context, documentId, revision)
        val (uses, ul) = usages.detect(context, documentId, revision)
        return SemanticSnapshot(documentId, revision, defs, uses, dl, ul)
    }
    /** Cursor is in usage.documentId. A completion query accepts only prefix-derived local records. */
    fun resolve(usage: SymbolResult, candidates: List<SymbolResult>, cursor: Int? = usage.start,
                threshold: Float = .8f, completionContext: Boolean = false): DefinitionResolution {
        require(intelligence.trained()) { "intelligence has no trained link examples" }
        require(threshold in 0f..1f && candidates.size <= 512)
        require(candidates.map { Triple(it.documentId, it.revision, it.id) }.toSet().size == candidates.size)
        if (completionContext) {
            require(cursor != null && usage.end <= cursor) { "usage is after completion cursor" }
            require(usage.contextEnd != null && usage.contextEnd <= cursor) { "usage was not detected from prefix-only context" }
            require(candidates.none { it.documentId == usage.documentId &&
                (it.end > cursor || it.contextEnd == null || it.contextEnd > cursor) }) { "candidate was not detected from prefix-only context" }
        }
        val scores = FloatArray(candidates.size + 1) { index ->
            val features = features(usage, candidates.getOrNull(index), cursor)
            val hidden = linear(features, hw, hb).map { tanh(it) }.toFloatArray()
            linear(hidden, sw, sb)[0]
        }
        val probs = softmax(scores); val best = argmax(probs)
        val target = if (best < candidates.size && probs[best] >= threshold) candidates[best] else null
        return DefinitionResolution(usage, target, probs[best], probs)
    }
    companion object {
        val KINDS = listOf("unknown", "variable", "function", "type", "field", "parameter", "module", "constant")
        val FEATURE_NAMES = listOf("null", "sameDocument", "sameName", "sameKind", "usageConfidence",
            "definitionConfidence", "definitionBeforeUsage", "signedDistance") +
            listOf("usage", "definition").flatMap { role -> KINDS.map { "$role" + "Kind:$it" } } +
            listOf("queryPresent", "usagePosition", "definitionPosition", "queryPosition", "usageToQuery",
                   "absoluteUsageToQuery", "definitionToQuery", "absoluteDefinitionToQuery")
        fun features(u: SymbolResult, d: SymbolResult?, cursor: Int? = null): FloatArray {
            require(cursor == null || cursor >= 0)
            val same = d != null && u.documentId == d.documentId
            require(!same || u.revision == d!!.revision) { "stale definition revision" }
            fun flag(b: Boolean) = if (b) 1f else 0f
            fun bounded(p: Int) = p.toDouble().div(p.toDouble() + 4096.0).toFloat()
            fun distance(p: Int) = if (cursor != null) ((p.toDouble() - cursor) / 4096.0).coerceIn(-1.0, 1.0).toFloat() else 0f
            val uf = distance(u.start); val df = if (same) distance(d!!.start) else 0f
            return (listOf(flag(d == null), flag(same), flag(d != null && u.name == d.name),
                flag(d != null && u.kind == d.kind), u.confidence, d?.confidence ?: 0f,
                flag(same && d!!.start <= u.start), if (same) ((d!!.start.toDouble() - u.start) / 4096.0).coerceIn(-1.0, 1.0).toFloat() else 0f) +
                (0..7).map { flag(u.kind == it) } + (0..7).map { flag(d != null && d.kind == it) } +
                listOf(flag(cursor != null), bounded(u.start), if (same) bounded(d!!.start) else 0f,
                       cursor?.let { bounded(it) } ?: 0f, uf, abs(uf), df, abs(df))).toFloatArray()
        }
    }
}

File: training/semantic_models.py
"""Small independent detectors and a ranker over their public output records."""
from __future__ import annotations

import hashlib
from dataclasses import asdict, dataclass
from pathlib import Path

from .annotations import range_from, utf16_range_to_bytes
from .syl2_format import read_syl2, write_syl2
from .syl2_trainer import BOS, BYTE_VOCABULARY, SYMBOL_KIND_LABELS, _kind_index, _torch
from .registry import LANGUAGES, REGISTRY_HASH
from .semantic_context import SemanticContext

SCHEMA = "semantic-records-v1"
MAX_CANDIDATES = 512
BIO = ("none", "begin", "inside")
FEATURE_NAMES = ("null", "sameDocument", "sameName", "sameKind", "usageConfidence",
                 "definitionConfidence", "definitionBeforeUsage", "signedDistance") + tuple(
    f"{role}Kind:{kind}" for role in ("usage", "definition") for kind in SYMBOL_KIND_LABELS) + (
    "queryPresent", "usagePosition", "definitionPosition", "queryPosition",
    "usageToQuery", "absoluteUsageToQuery", "definitionToQuery", "absoluteDefinitionToQuery")


@dataclass(frozen=True)
class SymbolRecord:
    id: str
    document_id: str
    revision: str
    start: int
    end: int
    name: str
    kind: int
    confidence: float
    context_end: int | None = None
    context: SemanticContext | None = None

    def __post_init__(self):
        if not self.id or not self.document_id or not self.revision or not 0 <= self.start <= self.end:
            raise ValueError("invalid symbol identity or range")
        if not 0 <= self.kind < len(SYMBOL_KIND_LABELS) or not 0 <= self.confidence <= 1:
            raise ValueError("invalid symbol kind/confidence")
        if self.context_end is not None and self.context_end < self.end:
            raise ValueError("symbol is outside the analyzed context")


def make_detector(seed=17):
    torch, nn, _ = _torch()
    torch.manual_seed(seed)

    class Detector(nn.Module):
        def __init__(self):
            super().__init__()
            self.embedding = nn.Embedding(BYTE_VOCABULARY, 64, padding_idx=258)
            self.encoder = nn.GRU(64, 64, num_layers=2, bidirectional=True, batch_first=True)
            self.occurrence_head = nn.Linear(128, 3)
            self.kind_head = nn.Linear(128, len(SYMBOL_KIND_LABELS))
            self.language_head = nn.Linear(128, len(LANGUAGES))

        def forward(self, ids):
            hidden, _ = self.encoder(self.embedding(ids))
            return self.occurrence_head(hidden), self.kind_head(hidden), self.language_head(hidden.mean(1))

    return Detector()


def make_intelligence(seed=19):
    torch, nn, _ = _torch()
    torch.manual_seed(seed)

    class Intelligence(nn.Module):
        def __init__(self):
            super().__init__()
            self.hidden = nn.Linear(len(FEATURE_NAMES), 64)
            self.score = nn.Linear(64, 1)

        def forward(self, features):
            return self.score(torch.tanh(self.hidden(features))).squeeze(-1)

    return Intelligence()


def boundaries(text):
    """(UTF-8 byte offset, UTF-16 code-unit offset) at every code-point boundary."""
    result = [(0, 0)]
    byte = unit = 0
    for char in text:
        byte += len(char.encode("utf-8"))
        unit += 2 if ord(char) > 0xFFFF else 1
        result.append((byte, unit))
    return result


def windows(text, sequence_bytes):
    if sequence_bytes < 4:
        raise ValueError("semantic windows must fit a complete UTF-8 code point (>=4 bytes)")
    points = boundaries(text)
    start = 0
    while start < len(text):
        end = start + 1
        while end < len(text) and points[end + 1][0] - points[start][0] <= sequence_bytes:
            end += 1
        yield start, end, points[start][0], points[end][0]
        start = end


def detector_targets(text, annotation, role):
    """Partial annotation never implies that all other source bytes are negatives."""
    size = len(text.encode("utf-8"))
    complete = annotation.teacher.get("occurrenceCoverage") == "complete"
    tags, kinds = [0 if complete else -100] * size, [-100] * size
    own = annotation.definitions if role == "definitions" else annotation.usages
    other = annotation.usages if role == "definitions" else annotation.definitions
    for item in other:
        a, b = range_from(item, "nameRange" if role == "usages" else "range")
        a, b = utf16_range_to_bytes(text, a, b)
        tags[a:b] = [0] * (b - a)
    for item in own:
        a, b = range_from(item, "nameRange" if role == "definitions" else "range")
        a, b = utf16_range_to_bytes(text, a, b)
        if a == b:
            raise ValueError("symbol annotations must have nonempty ranges")
        tags[a:b] = [2] * (b - a)
        tags[a] = 1
        kinds[a:b] = [_kind_index(item.get("kind"))] * (b - a)
    return tags, kinds


def detect(model, text, document_id, revision, role, sequence_bytes=512):
    torch, _, _ = _torch()
    device = next(model.parameters()).device
    points = boundaries(text)
    raw = text.encode("utf-8")
    # Per-character BIO decisions keep UTF-16 spans out of surrogate pairs.
    decisions, probabilities, kind_scores, language_scores = [], [], [], []
    model.eval()
    with torch.inference_mode():
        for a, b, ba, bb in windows(text, sequence_bytes):
            tags, kinds, languages = model(torch.tensor([[BOS] + list(raw[ba:bb])], device=device))
            probs = tags[0, 1:].softmax(-1).cpu()
            kinds = kinds[0, 1:].cpu()
            language_scores.append((languages[0].softmax(-1).cpu(), b - a))
            for index in range(a, b):
                pos = points[index][0] - ba
                label = int(probs[pos].argmax())
                decisions.append(label)
                probabilities.append(float(probs[pos, label]))
                kind_scores.append(kinds[pos])
    records = []
    start = None
    for index in range(len(text) + 1):
        label = decisions[index] if index < len(text) else 0
        if start is not None and label != 2:
            a, b = points[start][1], points[index][1]
            records.append(SymbolRecord(f"{role}:{a}:{b}", document_id, revision, a, b,
                                        text[start:index], int(torch.stack(kind_scores[start:index]).mean(0).argmax()),
                                        sum(probabilities[start:index]) / (index - start), points[-1][1]))
            start = None
        if index < len(text) and label and start is None:
            start = index
    language = (sum(value * count for value, count in language_scores) / max(len(text), 1)).tolist() \
        if language_scores else [0.0] * len(LANGUAGES)
    return records, language


def pair_features(usage, definition, query_position=None):
    if query_position is not None and (type(query_position) is not int or query_position < 0):
        raise ValueError("query position must be a nonnegative UTF-16 offset")
    if definition is not None and usage.document_id == definition.document_id and usage.revision != definition.revision:
        raise ValueError("stale definition revision")
    same_doc = definition is not None and usage.document_id == definition.document_id
    distance = max(-1.0, min(1.0, (definition.start - usage.start) / 4096.0)) if same_doc else 0.0
    relative = lambda position: max(-1., min(1., (position - query_position) / 4096.)) if query_position is not None else 0.
    bounded = lambda position: position / (position + 4096.)
    u_distance = relative(usage.start)
    d_distance = relative(definition.start) if same_doc else 0.
    return [float(definition is None), float(same_doc),
            float(definition is not None and usage.name == definition.name),
            float(definition is not None and usage.kind == definition.kind), usage.confidence,
            definition.confidence if definition else 0.0,
            float(same_doc and definition.start <= usage.start), distance] + \
        [float(usage.kind == k) for k in range(len(SYMBOL_KIND_LABELS))] + \
        [float(definition is not None and definition.kind == k) for k in range(len(SYMBOL_KIND_LABELS))] + \
        [float(query_position is not None), bounded(usage.start), bounded(definition.start) if same_doc else 0.,
         bounded(query_position) if query_position is not None else 0.,
         u_distance, abs(u_distance), d_distance, abs(d_distance)]


def candidate_features(usage, definitions, query_position=None):
    if len(definitions) > MAX_CANDIDATES:
        raise ValueError("candidate context exceeds 512 definitions; narrow the query context")
    identities = [(d.document_id, d.revision, d.id) for d in definitions]
    if len(set(identities)) != len(identities):
        raise ValueError("duplicate definition candidate identity")
    return [pair_features(usage, definition, query_position) for definition in definitions] + [pair_features(usage, None, query_position)]


def rank(model, usage, definitions, threshold=0.8, query_position=None, completion_context=False):
    if not 0 <= threshold <= 1:
        raise ValueError("threshold must be in [0,1]")
    if completion_context and (query_position is None or usage.end > query_position or
                               usage.context_end is None or usage.context_end > query_position or
                               any(d.document_id == usage.document_id and
                                   (d.end > query_position or d.context_end is None or d.context_end > query_position)
                                   for d in definitions)):
        raise ValueError("completion context contains records after the cursor")
    torch, _, _ = _torch()
    model.eval()
    with torch.inference_mode():
        scores = model(torch.tensor(candidate_features(usage, definitions, query_position), dtype=torch.float32,
                                    device=next(model.parameters()).device)).softmax(0).cpu()
    best = int(scores.argmax())
    confidence = float(scores[best])
    target = definitions[best] if best < len(definitions) and confidence >= threshold else None
    return {"usage": asdict(usage), "definition": asdict(target) if target else None,
            "confidence": confidence, "status": "resolved" if target else "unresolved",
            "candidateProbabilities": scores.tolist()}


def export(path, model, task, *, sequence_bytes=512, trained_steps=0, dependencies=None):
    write_syl2(path, {"schema": SCHEMA, "taskId": task, "registryHash": REGISTRY_HASH,
                     "languages": list(LANGUAGES), "sequenceBytes": sequence_bytes,
                     "bioLabels": list(BIO), "symbolKinds": list(SYMBOL_KIND_LABELS),
                     "featureNames": list(FEATURE_NAMES), "trainedSteps": trained_steps,
                     "status": "ANNOTATION_SUPERVISED" if trained_steps else "UNTRAINED",
                     "dependencies": dependencies or {},
                     "parameterCount": sum(p.numel() for p in model.parameters())}, model.state_dict())


def load(path, expected_task):
    if expected_task not in ("definitions", "usages", "code-intelligence"):
        raise ValueError("unknown semantic task")
    import numpy as np
    torch, _, _ = _torch()
    metadata, tensors = read_syl2(path)
    if metadata.get("schema") != SCHEMA or metadata.get("taskId") != expected_task:
        raise ValueError("incompatible semantic artifact")
    if metadata.get("registryHash") != REGISTRY_HASH or metadata.get("featureNames") != list(FEATURE_NAMES):
        raise ValueError("semantic registry or feature contract mismatch")
    model = make_intelligence() if expected_task == "code-intelligence" else make_detector()
    model.load_state_dict({name: torch.from_numpy(np.frombuffer(payload, dtype="<f4").copy().reshape(shape))
                           for name, (shape, payload) in tensors.items()}, strict=True)
    return model.eval(), metadata


def artifact_hash(path):
    return hashlib.sha256(Path(path).read_bytes()).hexdigest()
